numpyro/infer/mixed_hmc.py

```diff
@@ -0,0 +1,211 @@
+                *model_args, _gibbs_sites=z_gibbs, **model_kwargs)(z_hmc)
+
+        def update_discrete(idx, rng_key, hmc_state, z_discrete, ke_discrete):
+            # Algo 1, line 19: get a new discrete proposal
+            rng_key, z_discrete_new, pe_new, log_accept_ratio = self._discrete_proposal_fn(
+                rng_key, z_discrete, hmc_state.potential_energy,
+                partial(potential_fn, z_hmc=hmc_state.z), idx, self._support_sizes_flat[idx])
+            # Algo 1, line 20: depending on reject or refract, we will update
+            # the discrete variable and its corresponding kinetic energy. In case of
+            # refract, we will need to update the potential energy and its grad w.r.t. hmc_state.z
+            ke_discrete_i_new = ke_discrete[idx] + log_accept_ratio
+            grad_ = jacfwd if self.inner_kernel._forward_mode_differentiation else grad
+            z_discrete, pe, ke_discrete_i, z_grad = lax.cond(
+                ke_discrete_i_new > 0,
+                (z_discrete_new, pe_new, ke_discrete_i_new),
+                lambda vals: vals + (grad_(partial(potential_fn, vals[0]))(hmc_state.z),),
+                (z_discrete, hmc_state.potential_energy, ke_discrete[idx], hmc_state.z_grad),
+                identity)
+
+            ke_discrete = ops.index_update(ke_discrete, idx, ke_discrete_i)
+            hmc_state = hmc_state._replace(potential_energy=pe, z_grad=z_grad)
+            return rng_key, hmc_state, z_discrete, ke_discrete
+
+        def update_continuous(hmc_state, z_discrete):
+            model_kwargs_ = model_kwargs.copy()
+            model_kwargs_["_gibbs_sites"] = z_discrete
+            hmc_state_new = self.inner_kernel.sample(hmc_state, model_args, model_kwargs_)
+
+            # each time a sub-trajectory is performed, we need to reset i and adapt_state
+            # (we will only update them at the end of HMCGibbs step)
+            # For `num_steps`, we will record its cumulative sum for diagnostics
+            hmc_state = hmc_state_new._replace(i=hmc_state.i,
+                                               adapt_state=hmc_state.adapt_state,
+                                               num_steps=hmc_state.num_steps + hmc_state_new.num_steps)
+            return hmc_state
+
+        def body_fn(i, vals):
+            rng_key, hmc_state, z_discrete, ke_discrete, arrival_times = vals
+            idx = jnp.argmin(arrival_times)
+            # NB: length of each sub-trajectory is scaled from the current min(arrival_times)
+            # (see the note at total_time below)
+            trajectory_length = arrival_times[idx] * time_unit
+            arrival_times = arrival_times - arrival_times[idx]
+            arrival_times = ops.index_update(arrival_times, idx, 1.)
+
+            # this is a trick, so that in a sub-trajectory of HMC, we always accept the new proposal
+            pe = jnp.inf if self.inner_kernel._algo == "HMC" else hmc_state.potential_energy
```

Comment:
when does `self.inner_kernel._algo != "HMC"` ?

Reply:
I leave it to remind me that for NUTS, this trick is not required (and using `inf` in NUTS will give wrong results). I think we can relax the assertation `assert inner_kernel._algo == "HMC"` in the future, when some theory is verified.

Reply:
do you still want to keep this?